deckhand/engine/layering.py

```diff
@@ -0,0 +1,292 @@
+
+        Only documents with the same `schema` are allowed to be layered
+        together into a fully rendered document.
+    """
+
+    SUPPORTED_METHODS = ('merge', 'replace', 'delete')
+    LAYERING_POLICY_SCHEMA = 'deckhand/LayeringPolicy/v1'
+
+    def __init__(self, documents):
+        """Contructor for ``DocumentLayering``.
+
+        :param documents: List of YAML documents represented as dictionaries.
+        """
+        self.documents = [document.Document(d) for d in documents]
+        self._find_layering_policy()
+        self.layered_docs = self._calc_document_children()
+
+    def render(self):
+        """Perform layering on the set of `documents`.
+
+        Each concrete document will undergo layering according to the actions
+        defined by its `layeringDefinition`.
+
+        :returns: the list of rendered documents (does not include layering
+            policy document).
+        """
+        # ``rendered_data_by_layer`` agglomerates the set of changes across all
+        # actions across each layer for a specific document. Use copy
+        # to prevent global data from being updated referentially.
+        rendered_data_by_layer = {}
+
+        # NOTE(fmontei): ``global_docs`` represents the topmost documents in
+        # the system. It should probably be impossible for more than 1
+        # top-level doc to exist, but handle multiple for now.
+        global_docs = [doc for doc in self.layered_docs
+                       if doc.get_layer() == self.layer_order[0]]
+
+        for doc in global_docs:
+            layer_idx = self.layer_order.index(doc.get_layer())
+            rendered_data_by_layer[layer_idx] = doc.all_data
+
+            # Keep iterating as long as a child exists.
+            for child in doc.get_children(nested=True):
+
+                # Retrieve the most up-to-date rendered_data (by
+                # referencing the child's parent's data).
+                child_layer_idx = self.layer_order.index(child.get_layer())
+                rendered_data = rendered_data_by_layer[child_layer_idx - 1]
+
+                # Apply each action to the current document.
+                actions = child.get_actions()
+                for action in actions:
+                    rendered_data = self._apply_action(
+                        action, child.all_data, rendered_data)
+
+                # Update the actual document data if concrete.
+                if not child.is_abstract():
+                    self.layered_docs[self.layered_docs.index(child)][
+                        'data'] = rendered_data['data']
+
+                # Update ``rendered_data_by_layer`` for this layer so that
+                # children in deeper layers can reference the most up-to-date
+                # changes.
+                rendered_data_by_layer[child_layer_idx] = rendered_data
+
+            if 'children' in doc:
+                del doc['children']
+
+        return [d.all_data for d in self.layered_docs]
```

Comment:
This'll be a little easier to understand if it's ``d.to_dict() for d in self.layered_docs``.